channelserver/src/meta.rs

```diff
@@ -0,0 +1,219 @@
+use std::collections::BTreeMap;
+
+use actix_web::{http, HttpRequest};
+use maxminddb::{self, geoip2::City};
+
+use session::WsChannelSessionState;
+
+// Sender meta data, drawn from the HTTP Headers of the connection counterpart.
+#[derive(Serialize, Debug, Clone)]
+pub struct SenderData {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub ua: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub addr: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub city: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub region: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub country: Option<String>,
+}
+
+impl Default for SenderData {
+    fn default() -> Self {
+        Self {
+            ua: None,
+            addr: None,
+            city: None,
+            region: None,
+            country: None,
+        }
+    }
+}
+
+// Parse the Accept-Language header to get the list of preferred languages.
+// We default to "en" because of well-established Anglo-biases.
+fn preferred_language(alheader: String) -> Vec<String> {
+    let default_lang = String::from("en");
+    let mut lang_tree: BTreeMap<String, String> = BTreeMap::new();
+    let mut i = 0;
+    alheader.split(",").for_each(|l| {
+        if l.contains(";") {
+            let weight: Vec<&str> = l.split(";").collect();
+            let lang = weight[0].to_ascii_lowercase();
+            let pref = weight[1].to_ascii_lowercase();
+            lang_tree.insert(String::from(pref), String::from(lang));
+        } else {
+            lang_tree.insert(
+                format!("q=1.{:02}", i),
+                String::from(l.to_ascii_lowercase()),
+            );
+            i += 1;
+        }
+    });
+    let mut langs: Vec<String> = lang_tree.values().map(|l| l.to_owned()).collect();
+    langs.reverse();
+    langs.push(default_lang);
+    langs
+}
+
+// Return the element that most closely matches the preferred language.
+// This rounds up from the dialect if possible.
+fn get_preferred_language_element(
+    langs: &Vec<String>,
+    elements: BTreeMap<String, String>,
+) -> Option<String> {
+    for lang in langs.clone() {
+        if elements.contains_key(&lang) {
+            //unwrap is safe-ish because we've checked that the key exists.
+            return Some(elements.get(lang.as_str()).unwrap().to_string());
+        }
+        if lang.contains("-") {
+            let (lang, _) = lang.split_at(2);
+            if elements.contains_key(lang) {
+                return Some(elements.get(lang).unwrap().to_string());
+            }
+        }
+    }
+    None
```

Comment:
Do we expect to ever return `None` from this function?  IIUC the `langs` vec will always contain `"en"` as a fallback option, and I would expect the maxmind db to always have an English string available.  Should we consider erroring out noisily here rather than not returning anything?

Reply:
I'm actually tempted to make the "en" fallback be "*". It's quite possible for the langs to return back nothing (Say, if the user wants only "wa" and "eo"), thus the fallback. Much like the `iploc` call, there may be a fair bit of noise in the logs from these. Again, I'll look at adding a metric at least.